File: python/dp/ClimbingStairs.py
'''
You are climbing a staircase. It takes n steps to reach the top.
Each time you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?

Input: n = 2
Output: 2
Explanation: There are two ways to climb to the top.
1. 1 step + 1 step
2. 2 steps

Input: n = 3
Output: 3
Explanation: There are three ways to climb to the top.
1. 1 step + 1 step + 1 step
2. 1 step + 2 steps
3. 2 steps + 1 step
'''

class ClimbingStairs:
    """
    Time Complexity: O(N)
    Space Complexity: O(1)
    """
    def climbStairs(self, n: int) -> int:
        if n <= 2:
            return n
        
        # Base cases:
        # one: ways to reach step n-1
        # two: ways to reach step n-2
        one, two = 1, 1
        
        for i in range(n - 1):
            temp = one
            one = one + two
            two = temp
            
        return one

if __name__ == "__main__":
    solution = ClimbingStairs()
    
    # Test Case 1
    n1 = 2
    print(f"Steps: {n1}, Ways: {solution.climbStairs(n1)}") # Expected: 2

    # Test Case 2
    n2 = 3
    print(f"Steps: {n2}, Ways: {solution.climbStairs(n2)}") # Expected: 3
    
    # Test Case 3
    n3 = 5
    print(f"Steps: {n3}, Ways: {solution.climbStairs(n3)}") # Expected: 8

File: src/main/java/com/naresh/a_dsalgo/dp/dp1/ClimbingStairs.java
package com.naresh.a_dsalgo.dp.dp1;

/**
 * Problem: Climbing Stairs
 * Description: You are climbing a staircase. It takes n steps to reach the top. Each time you can either climb 1 or 2 steps.
 * In how many distinct ways can you climb to the top?
 */
public class ClimbingStairs {
    /**
     * Algorithm: Use iterative DP with two variables to track ways to reach the previous two steps.
     */
    public int climbStairs(int n) {
        // Pattern: DP (Space Optimized) | Time: O(n), Space: O(1)
        if (n <= 2) return n;
        var first = 1; var second = 2;
        for (int i = 3; i <= n; i++) {
            var third = first + second;
            first = second;
            second = third;
        }
        return second;
    }

    public static void main(String[] args) {
        var n = 5;
        var result = new ClimbingStairs().climbStairs(n);
        System.out.println("Ways to climb " + n + ": " + result);
    }

    /*
     * Dry Run:
     * Input: n = 5
     *
     * 1. Initialization:
     *    first = 1, second = 2
     *
     * 2. Loop Iterations:
     *    i = 3: third = 1 + 2 = 3 -> first = 2, second = 3
     *    i = 4: third = 2 + 3 = 5 -> first = 3, second = 5
     *    i = 5: third = 3 + 5 = 8 -> first = 5, second = 8
     *
     * 3. Result: 8
     */
    // FAANG Tip: This is essentially the Fibonacci sequence. O(1) space is expected for this problem.
}
